rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+With proxy enabled, a client may now ask a server to open connections to arbitrary hosts/ports, which may affect external services, not unlike open access proxies. An SMP proxy, by the virtue of being app-specific, can enforce more guardrails, limiting problems from malicious clients.
+
+- Failed `server` requests may incur a penalty/backpressure to protect external services and proxy.
+- Forwarding is blocked until a reply comes back to protect proxy.
+- Minimal client response time at ~2 seconds to protect client.
+- Forwarding requests are pulled from client queues and handled one by one to protect relay.
+- Fair client scheduling towards a relay to protect clients from each other.
+
+Each proxy-to-relay (P2R) connection aggregates traffic from multiple clients and protected by a
+
+A proxy runs a barebones SMP protocol towards each active relay, fronted by an `MVar` which, unlike other concurrency primitives, guarantee fairness. With it, each client "forwarding" worker takes a unit of work and tries to submit it to a relay worker.
+
+A client should not be able to inspect/probe relay connections and distinguish a fresh connection from an existing one. Thus each client gets a `server_id` map with ids distinct from the proxy internal relay map. This also precludes accidental coupling between implementation and protocol format with its (quite severe) limits.
+
+### Relay server
+
+As with notification subscriber "clients" a proxy is just an another client with a special set of commands for setup and forwarding.
+The usual SMP precautions apply.
+
+## Failure modes
+
+### Delivery problems
+
+A delivery latency may increase due to:
+- Slow client/proxy handshake.
+- Slow proxy/relay handshake for new servers.
+- Long time waiting for its turn to send due to lots of active clients.
+- Long time to deliver to relay due to overloaded server.
+This can be mitigated by proxy pool.
+A pool may be monitoring proxy latency in general and towards a specific server, then route accordingly.
+
+A message non-delivery may happen due to:
+- Proxy misconfiguration. Report error to user.
+- Relay stopped operating or proxy lost its connection. Signalled to client by proxy `ERR`. Retry later.
+- Proxy stopped operating or client lost its connection. Signalled by connection error. Retry with another proxy or later.
+
+A duplicate delivery may happen due to:
+- Client used multiple routes to deliver (e.g. because of apparent non-delivery).
+- Proxy replaying messages (maliciously or by misconfiguration).
+This is mitigated client side by deduplication. And, optionally, relay side, by using bloom filter with message DH keys.
+
+### Connectivity problems
+
+A proxy may lose connection to a relay.
+The protocol lacks proxy notifications, so a client may only be informed of the failure state when it tries to forward a message.
+If a client tries to send to a dropped connection it gets an `ERR` and must de-associate the server from the proxy.
```

Comment:
unclear. Implies some kind of association that was never introduced. What is dropped connection? How is it different from any other sending error?